cirq/atoms/atom_devices.py

```diff
@@ -0,0 +1,402 @@
+
+        # All qubits the operation acts on must be on the device
+        for q in operation.qubits:
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+
+        if isinstance(operation.gate, (ops.MeasurementGate, ops.IdentityGate)):
+            return
+
+        # Verify that a controlled gate operation is valid
+        if isinstance(operation, ops.GateOperation):
+            if len(operation.qubits) > self._max_parallel_c:
+                raise ValueError("Too many qubits acted on in parallel by a "
+                                 "controlled gate operation")
+            if len(operation.qubits) > 1:
+                for p in operation.qubits:
+                    for q in operation.qubits:
+                        distance = sqrt((p.row - q.row)**2 + (p.col - q.col)**2)
+                        if distance > self._control_radius:
+                            raise ValueError("Qubits {!r}, {!r} are too "
+                                             "far away".format(p, q))
+            return
+
+        # Verify that a valid number of Z gates are applied in parallel
+        if isinstance(operation.gate, ops.ZPowGate):
+            if len(operation.qubits) > self._max_parallel_z:
+                raise ValueError("Too many Z gates in parallel")
+
+        # Verify that a valid number of XY gates are applied in parallel
+        if isinstance(operation.gate, (ops.XPowGate,
+                                       ops.YPowGate,
+                                       ops.PhasedXPowGate)):
+            if (len(operation.qubits) > self._max_parallel_xy and
+                    len(operation.qubits) != len(self.qubits)):
+                raise ValueError("Bad number of XY gates in parallel")
+
+    def validate_moment(self, moment: ops.Moment):
+        """
+        Raises an error if the given moment is invalid on this device
+
+        Args:
+            moment: The moment to validate
+
+        Raises:
+            ValueError: If the given moment is invalid
+        """
+        super().validate_moment(moment)
+
+        operations = moment.operations
+        num_parallel_z = 0
+        num_parallel_xy = 0
+        num_parallel_controlled_qubits = 0
+        hasMeasurement = False
+
+        # Count the number of each type of gate occuring in this moment.
+        # Also verify that gates of the same type are equal
+        for op in operations:
+
+            if isinstance(op.gate, ops.ZPowGate):
+                if num_parallel_z == 0:
+                    refZgate = op.gate
+                else:
+                    if refZgate != op.gate:
+                        raise ValueError("Non-identical Parallel Z gates")
+                num_parallel_z += len(op.qubits)
+
+            if isinstance(op.gate, (ops.XPowGate,
+                                    ops.YPowGate,
+                                    ops.PhasedXPowGate)):
+                if num_parallel_xy == 0:
```

Comment:
You can do Zs in parallel with XYs?

Reply:
They use different lasers so they have independent constraints.